### PR TITLE
Add pairing implementation

### PR DESCRIPTION
**Pull request by gudgl**
## User Story
As a supporter, I want to be able to guide the person who needs my help through the process of starting the agent. This must be possible on a phone call. 
1. I want to create a new session on my Plexus server.
2. The plexus server returns a short pairing code like "ABC123" that can be passed on by phone.
3. The person seeking help goes to https://pc-assist.app and downloads the PC Assist app (subject of this task) and download the version for the used OS.
4. After starting the app, the person entered the pairing code.
5. The app shows some confirmation message like "Do you really want to grant access to <HELPER>".
5. After confirming, the session starts. 

## Preconditions
A simple PHP/HTML page that serves the downloads of the PC Assist app will be created soon. (not subject of this task)

#### Config extension
Extend the plexus.config with optional settings like
```
# Publish all sessions on a pairing service
PLEXUS_PAIRING_URL=https://pc-assist.app/pairing/
# Pairing codes will expire after N seconds.
PLEXUS_PAIRING_TTL=600
# The pc-assist app can show your company name before the sessions start
PLEXUS_COMPANY_NAME="The great PC support"
# The pc-assist app can load your logo from a public URL
PLEXUS_COMPANY_LOGO="https://img.icons8.com/external-wanicon-lineal-color-wanicon/64/000000/external-help-friendship-wanicon-lineal-color-wanicon.png"
```
#### Session creation
If the `PLEXUS_PAIRING_URL` is given
1. The plexus server must accept the two extra fields `supporter_name` and `supporter_avatar` on session creation. 
For example:
    ```
    curl -ks https://localhost:8080/session \
          -u support:1234 \
          -F id="helping-joe" \
          -F ttl=3600 \
          -F username=admin \
          -F password=foobaz \
          -F supporter_name=Thorsten \
          -F supporter_avatar="https://avatars.githubusercontent.com/u/16852023?v=4"
    ```
1. On session creation, the Plexus server must send a json POST request to `{PLEXUS_PAIRING_URL}` for example https://pc-assist.app/sessions/ containing the base URL of the new pairing endpoint of the Plexus server (see below).
    ```
    curl 'https://pc-assist.app/pairing/' \
    --header 'Content-Type: application/json' \
    --data-raw '{"url":"http://plexus.example.com/pairing"}'
    ```
    The above URL will return pairing data as follows. 
    ```json
    {
        "success": true,
        "code": "7auu83",
        "pairing_url": "https://pc-assist.app/pairing/7auu83",
        "redirect_url": "http://plexus.example.com/pairing/7auu83"
    }
    ```` 
    If you request the `pairing_url` you will get a 302 redirect to the `redirect_url`. 

1. if the pairing code has been retrieved successfully, it must be included into the response of the session creation, for example.
    ```json
    {
      "ID": "helping-joe",
      "SessionURL": "https://localhost:8080/session/helping-joe",
      "AgentMSH": "https://localhost:8080/config/helping-joe:xddRGfuIOaqwBxVIWrnp",
      "AgentConfig": {
        "ServerID": "9DA17836FD0BA3193ED52896929FD021990EBA4234ED56A6057115B7C53D24F58284E34954619CAECC131FC8BE82B9EE",
        "MeshName": "plexus/helping-joe/bGVnI",
        "MeshType": 2,
        "MeshID": "mesh//i8bgwSqhUVS5ttAYX5VCqSR2dxPY@5iSLv5p1jFJG5TJKYV91PaMoTf0AHSj@Eqi",
        "MeshIDHex": "0x8BC6E0C12AA15154B9B6D0185F9542A924767713D8FB98922EFE69D631491B94C929857DD4F68CA137F40074A3F84AA2",
        "MeshServer": "wss://localhost:8080/agent/helping-joe:xddRGfuIOaqwBxVIWrnp"
      },
      "ExpiresAt": "2021-09-12T14:31:40.830231373+02:00",
      "PairingCode": "7auu83",
      "PairingUrl":"https://pc-assist.app/pairing/7auu83"
    }
    ```
#### Pairing creation
Now it's time to provide a pairing URL on Plexus.
If the request to the public pairing service on https://pc-assist.app has returned a pairing code and a redirect URL, store the code in the memory of the plexus server and make the redirect URL available. The PC Assist desktop app will connect there.
With the above example, the plexus server must make the URL `GET https://localhost:8080/pairing/7auu83` available. It must return a json object like
```json
{
  "AgentMSH": "https://localhost:8080/config/helping-joe:xddRGfuIOaqwBxVIWrnp",
  "ExpiresAt": "2021-09-12T14:31:40.830231373+02:00",
  "CompanyName": "The great PC support",
  "CompanyLogo": "https://img.icons8.com/external-wanicon-lineal-color-wanicon/64/000000/external-help-friendship-wanicon-lineal-color-wanicon.png",
  "SupporterName":"Thorsten",
  "SupporterAvatar":"https://avatars.githubusercontent.com/u/16852023?v=4"
}
```

### CODE REVIEW

File: handler/create.go [hunk 1]
```diff
@@ -88,6 +93,34 @@ func (h *Handler) CreateSession(rw http.ResponseWriter, r *http.Request) {
 			MeshType:   2,
 		},
 	}
+
+	if h.pcfg.PairingURL != "" {
+		if supName == "" || supAvatar == "" {
+			api.WriteJSONResponse(rw, http.StatusBadRequest, map[string]interface{}{
+				"message": "You need to provide supporter_name and supporter_avatar for pairing",
+			})
```
Comment (jmattheis) [inline handler/create.go:101]: Use api.WriteJSONError here.

File: handler/create.go [hunk 2]
```diff
@@ -88,6 +93,34 @@ func (h *Handler) CreateSession(rw http.ResponseWriter, r *http.Request) {
...
+			return
+		}
+		pr, err := pairing.Pair(r.Context(), h.pcfg.PairingURL, &pairing.Request{
+			Url: "",
+		})
+		if err != nil {
+			api.WriteJSONResponse(rw, http.StatusInternalServerError, map[string]interface{}{
```
Comment (jmattheis) [inline handler/create.go:108]: Use api.WriteJSONError here.

File: handler/create.go [hunk 3]
```diff
@@ -88,6 +93,34 @@ func (h *Handler) CreateSession(rw http.ResponseWriter, r *http.Request) {
...
+
+	if h.pcfg.PairingURL != "" {
+		if supName == "" || supAvatar == "" {
+			api.WriteJSONResponse(rw, http.StatusBadRequest, map[string]interface{}{
+				"message": "You need to provide supporter_name and supporter_avatar for pairing",
+			})
+			return
+		}
+		pr, err := pairing.Pair(r.Context(), h.pcfg.PairingURL, &pairing.Request{
+			Url: "",
```
Comment (jmattheis) [inline handler/create.go:105]: This shouldn't be an empty string right?

File: pairing/pair.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	}
+
+	defer response.Body.Close()
+	jsonResponse, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		log.Error().Msg(string(jsonResponse))
```
Comment (jmattheis) [inline pairing/pair.go:55]: this code must use the logger that is defined in handler., it would be good, to also log the status code.

File: pairing/pair.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	if err != nil {
+		return nil, err
+	}
+	client := &http.Client{
+		Timeout: defaultTimeout,
+	}
+
+	response, err := ctxhttp.Post(ctx, client, url, contentType, bytes.NewBuffer(jsonRequest))
+	if err != nil {
+		return nil, err
```
Comment (jmattheis) [inline pairing/pair.go:45]: Please add some context to the each error in this method. Example:
```suggestion
		return nil, fmt.Errorf("could not post request to %s: %s", url, err)
```

File: handler/create.go (hunk 2 above)
Comment (jmattheis) [inline handler/create.go:108]: This should be a http.StatusBadGateway

File: handler/handler.go [hunk 6]
```diff
@@ -45,16 +48,20 @@ func Register(r *mux.Router, opt *Options) {
...
 }
 
 type Handler struct {
 	log                logger.Logger
-	cfg                *control.Config
+	ccfg               *control.Config
+	pcfg               *pairing.Config
 	auth               AuthChecker
 	lock               sync.RWMutex
+	codeLock           sync.RWMutex
```
Comment (jmattheis) [inline handler/handler.go:61]: Use the existing lock instead of codeLock, in almost all cases you also need the normal session lock, thus, no need for two locks. Having to locks just adds the possibility for deadlocks.

File: handler/pair.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+// @Param code path string true "code"
+// @Security BasicAuth
+// @Success 200 {object} api.URLResponse
+// @Failure 401 {object} api.Error
+// @Failure 404 {object} api.Error
+// @Router /session/{code} [get]
+func (h *Handler) Pair(rw http.ResponseWriter, r *http.Request) {
+	h.codeLock.Lock()
+	id, ok := h.codes[mux.Vars(r)["id"]]
+	h.codeLock.Unlock()
```
Comment (jmattheis) [inline handler/pair.go:26]: checkSessionAuthentication requires a lock of h.lock.
```suggestion
	h.lock.Lock()
	defer h.lock.Unlock()
	id, ok := h.codes[mux.Vars(r)["id"]]
```

File: pairing/pair.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+	if response.StatusCode != http.StatusOK {
+		log.Error().Msg(string(jsonResponse))
+		return nil, ErrUnableToPair
+	}
+
+	resp := Response{}
+	err = json.Unmarshal(jsonResponse, &resp)
+	if err != nil {
+		return nil, err
+	}
```
Comment (jmattheis) [inline pairing/pair.go:63]: Add some debug log when the pairing succeeds.

File: pairing/config.go [hunk 9]
```diff
@@ -0,0 +1,8 @@
+package pairing
+
+type Config struct {
+	PairingURL  string `split_words:"true"`
```
Comment (jmattheis) [inline pairing/config.go:4]: No need for the split_words in the tag here.

**Review by jmattheis — CHANGES_REQUESTED** [review]

File: handler/create.go [hunk 10]
```diff
@@ -95,30 +95,28 @@ func (h *Handler) CreateSession(rw http.ResponseWriter, r *http.Request) {
...
 		})
 		if err != nil {
-			api.WriteJSONResponse(rw, http.StatusInternalServerError, map[string]interface{}{
-				"message": "Unable to create session, failed to pair",
-			})
+			api.WriteJSONError(rw, http.StatusBadGateway, "Unable to create session, failed to pair")
 			return
 		}
 
+		h.log.Infof("pairing succeeded")
```
Comment (jmattheis) [inline handler/create.go:111]: Use Debugf here.

File: handler/create.go [hunk 11]
```diff
@@ -95,30 +95,28 @@ func (h *Handler) CreateSession(rw http.ResponseWriter, r *http.Request) {
 	}
 
 	if h.pcfg.PairingURL != "" {
+		h.log.Infof("pairing...")
```
Comment (jmattheis) [inline handler/create.go:97]: use Debugf here.

File: handler/pcpair.go [hunk 12]
```diff
@@ -31,35 +31,36 @@ type Response struct {
...
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("post failed: %w", err)
 	}
 
 	defer response.Body.Close()
 	jsonResponse, err := ioutil.ReadAll(response.Body)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("reading body failed: %w", err)
```
Comment (jmattheis) [inline handler/pcpair.go:48]: Add the http status here too.

File: handler/create.go [hunk 13]
```diff
@@ -95,30 +95,28 @@ func (h *Handler) CreateSession(rw http.ResponseWriter, r *http.Request) {
...
-		pr, err := pairing.Pair(r.Context(), h.pcfg.PairingURL, &pairing.Request{
-			Url: "",
+		pr, err := pcpairing.Pair(r.Context(), h.pcfg.PairingURL, &pcpairing.Request{
+			Url: h.pcfg.PairingURL,
 		})
 		if err != nil {
-			api.WriteJSONResponse(rw, http.StatusInternalServerError, map[string]interface{}{
-				"message": "Unable to create session, failed to pair",
-			})
+			api.WriteJSONError(rw, http.StatusBadGateway, "Unable to create session, failed to pair")
```
Comment (jmattheis) [inline handler/create.go:107]: maybe add the `err` as string to the error message.

File: config/config.go [hunk 14]
```diff
@@ -11,6 +11,7 @@ import (
 	"github.com/rs/zerolog"
 
 	"github.com/cloudradar-monitoring/plexus/control"
+	"github.com/cloudradar-monitoring/plexus/pcpairing"
```
Comment (jmattheis) [inline config/config.go:14]: Why did you rename this to pcpairing? IMO pairing was better, as you could possibly use the paring functionality on a server.

File: handler/create.go [hunk 15]
```diff
@@ -88,6 +92,32 @@ func (h *Handler) CreateSession(rw http.ResponseWriter, r *http.Request) {
...
+
+	if h.pcfg.PairingURL != "" {
+		h.log.Infof("pairing...")
+
+		if supName == "" || supAvatar == "" {
+			api.WriteJSONError(rw, http.StatusBadRequest, "You need to provide supporter_name and supporter_avatar for pairing")
+			return
+		}
+		pr, err := h.pcPair(r.Context(), h.pcfg.PairingURL, &Request{
+			Url: h.pcfg.PairingURL,
```
Comment (jmattheis) [inline handler/create.go:104]: This should be the URL of the current plexus instance.

File: handler/pair.go [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+// @Param code path string true "code"
+// @Security BasicAuth
+// @Success 200 {object} api.URLResponse
+// @Failure 401 {object} api.Error
+// @Failure 404 {object} api.Error
+// @Router /session/{code} [get]
+func (h *Handler) Pair(rw http.ResponseWriter, r *http.Request) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+	id, ok := h.codes[mux.Vars(r)["id"]]
```
Comment (jmattheis) [inline handler/pair.go:26]: I think it would be better iterating over the session map and finding a session that has the pairing code. This extra map just adds complexity. The session map already will be cleaned accordingly, the codes map will just grow endlessly.

File: handler/pair.go [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+		})
+		return
+	}
+
+	session, ok := h.checkSessionAuthentication(rw, r, id)
+	if !ok {
+		return
+	}
+
+	api.WriteJSONResponse(rw, http.StatusNotFound, api.PairedSession{
```
Comment (jmattheis) [inline handler/pair.go:39]: this should be status ok.

File: handler/pair.go [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+// @Success 200 {object} api.URLResponse
+// @Failure 401 {object} api.Error
+// @Failure 404 {object} api.Error
+// @Router /session/{code} [get]
+func (h *Handler) Pair(rw http.ResponseWriter, r *http.Request) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+	id, ok := h.codes[mux.Vars(r)["id"]]
+	if !ok {
+		api.WriteJSONResponse(rw, http.StatusNotFound, map[string]interface{}{
```
Comment (jmattheis) [inline handler/pair.go:28]: use api.WriteJsonError here.

File: handler/create.go (hunk 11 above)
Comment (jmattheis) [reply handler/create.go:97]: log the pairing url here. and log the pairing code on successful pairing.

File: handler/pair.go [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+// Pair godoc
+// @Summary Gets the pairing code info.
+// @Tags session
+// @Produce application/json
+// @Param code path string true "code"
+// @Security BasicAuth
+// @Success 200 {object} api.URLResponse
+// @Failure 401 {object} api.Error
+// @Failure 404 {object} api.Error
+// @Router /session/{code} [get]
```
Comment (jmattheis) [inline handler/pair.go:22]:
```suggestion
// @Router /pairing/{code} [get]
```

**Review by jmattheis — CHANGES_REQUESTED** [review]

File: handler/pair.go (hunk 16 above)
Comment (gudgl) [reply handler/pair.go:26]: When deleting sessions we can delete the code as well.
But if you think iterating is better okay ill change it.

File: handler/pair.go (hunk 17 above)
Comment (gudgl) [reply handler/pair.go:39]: why status ok?

Comment (jmattheis) [reply handler/pair.go:37]: Because the request status should be a success and not a client error like not found?

Comment (gudgl) [reply handler/pair.go:37]: sorry my bad I taught you wanted the other code on 28 line to be status ok I didn't even look at this code

File: handler/pair.go [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+	id, ok := h.getSessionID(mux.Vars(r)["code"])
+	if !ok {
+		api.WriteJSONError(rw, http.StatusNotFound, "Code not found")
+		return
+	}
+
+	session, ok := h.checkSessionAuthentication(rw, r, id)
+	if !ok {
+		return
+	}
```
Comment (jmattheis) [inline handler/pair.go:35]: I think the pairing shouldn't require session authentication. As this makes creating a connection with the user more difficult. See issue description:
> , I want to be able to guide the person who needs my help through the process of starting the agent. This must be possible on a phone call.

I'd say only the pairing code should be requried.

**Review by jmattheis — CHANGES_REQUESTED** [review]

File: handler/pair.go [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+	if !ok {
+		api.WriteJSONError(rw, http.StatusNotFound, "Code not found")
+		return
+	}
+
+	session, ok := h.sessions[id]
+	if !ok {
+		api.WriteJSONError(rw, http.StatusNotFound, fmt.Sprintf("Session paired to code %s doesn't exist", code))
+		return
+	}
```
Comment (jmattheis) [inline handler/pair.go:43]: In getSessionID you already have the session and only return the session id to then lookup the sessionid in the session map. How about renaming getSessionID to getSessionIDByPairingCode and returning the session instance and a boolean in there.
```
func (h *Handler) getSessionIDByPairingCode(code string) (*Session, bool) {
```

Comment (gudgl) [reply handler/pair.go:43]: good catch

File: handler/create.go [hunk 22]
```diff
@@ -107,9 +141,70 @@ func (h *Handler) CreateSession(rw http.ResponseWriter, r *http.Request) {
...
+
+func (h *Handler) pcPair(ctx context.Context, supName, supAvatar string, session *Session) error {
+	h.log.Debugf("pairing to %s ...", h.pcfg.PairingURL)
+
+	if supName == "" || supAvatar == "" {
+		return ErrMissingSupporterInfo
+	}
+
+	jsonRequest, _ := json.Marshal(&Request{
+		URL: fmt.Sprintf("https://%s%s/pairing", h.pcfg.ServerAddress, h.prefix),
```
Comment (jmattheis) [inline handler/create.go:172]: The server address can be something like 0.0.0.0:8080, this is not a valid url that is can be connected to. You need to use the host of the http request to find the originating host of the request. Similarly it is done when returning the session url in the create session request.
```bash
jm@dyne ~ λ https --verify=no -a admin:admin POST ':8080/session' id==helping-joe ttl==3600 supporter_name==help supporter_avatar==k|jq
{
  "ID": "helping-joe",
  "SessionURL": "https://localhost:8080/session/helping-joe",
  "AgentMSH": "https://localhost:8080/config/helping-joe:VFrx3t6Vo89Cb2BdADGZ",
  "PairingCode": "b6ydr1",
  "PairingURL": "https://pc-assist.app/pairing/b6ydr1",
  "AgentConfig": {
    "ServerID": "27C73371BAD57819ABBC8977C57619C382FCF8C846E0DBCAFA77AB72F6D454AA5A0B8CF41151DCBF24D29B6717B2EEAB",
    "MeshName": "plexus/helping-joe/9bnMI",
    "MeshType": 2,
    "MeshID": "mesh/control/TwcFgmnFJBi9wiNhVcPEuMjZal1SS@z9o5gSvehVVDkTABgjD2@ebyroN6ZNCFfI",
    "MeshIDHex": "0x4F07058269C52418BDC2236155C3C4B8C8D96A5D524BECFDA39812BDE8555439130018230F6F9E6F2AE837A64D0857C8",
    "MeshServer": "wss://localhost:8080/agent/helping-joe:VFrx3t6Vo89Cb2BdADGZ"
  },
  "ExpiresAt": "2021-11-16T21:36:21.926363494+01:00"
}
jm@dyne ~ λ http https://pc-assist.app/pairing/b6ydr1
HTTP/1.1 302 Found
*.js: max-age=31536000
Cache-Control: no-cache, private
Content-Length: 390
Content-Type: text/html; charset=UTF-8
Date: Tue, 16 Nov 2021 19:36:33 GMT
Location: https://0.0.0.0:8080//pairing/b6ydr1
Server: Caddy
Status: 302 Found
```

File: pairing/config.go [hunk 23]
```diff
@@ -0,0 +1,9 @@
+package pairing
+
+type Config struct {
```
Comment (jmattheis) [inline pairing/config.go:3]: The paring config should be moved to handler.Options no need for a separate package with only one config struct.

File: config/config.go [hunk 24]
```diff
@@ -58,8 +57,16 @@ func (s *Config) AsControlConfig() *control.Config {
 	}
 }
 
-func (s *Config) AsPairingConfig() *pairing.Config {
-	return &pairing.Config{
+type PairingConfig struct {
+	ServerAddress string
```
Comment (jmattheis) [inline config/config.go:61]: Can be removed.

**Review by jmattheis — APPROVED** [review]
One little remark, then you can merge & rebase it.
